Add image gallery component

joshuatf opened the pull request:
### All Submissions:

-   [x] Have you followed the [WooCommerce Contributing guideline](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md)?
-   [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
-   [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change?

<!-- Mark completed items with an [x] -->

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

This PR adds the image gallery component.  In order to do this it also:
* Refactors the `Sortable` component to allow sorting vertically
* Extracts the `ListItem` component to reduce opinionated styling of sortable items

Note that the toolbar for the Image Gallery will be a follow-up in another PR.  This one focuses on base gallery styling and sortability.


<img width="375" alt="Screen Shot 2022-08-23 at 12 46 37 PM" src="https://user-images.githubusercontent.com/10561050/186252352-a2e6ca23-3286-486a-a138-6509e6874eca.png">
<img width="340" alt="Screen Shot 2022-08-23 at 12 34 27 PM" src="https://user-images.githubusercontent.com/10561050/186252355-af7f72d6-3e7e-4d52-aaf6-bf4fdf2f7bc3.png">
<img width="673" alt="Screen Shot 2022-08-23 at 12 34 21 PM" src="https://user-images.githubusercontent.com/10561050/186252359-22e20aaa-9886-40bd-a4f4-8555556bfdaa.png">

### How to test the changes in this Pull Request:

1. `npm run storybook`
2. View the `Sortable`, `ImageGallery`, and `ListItem` stories
3. Check that things work as expected

### Other information:

-   [x] Have you added an explanation of what your changes do and why you'd like us to include them?
-   [x] Have you written new tests for your changes, as applicable?
-   [x] Have you successfully run tests with your changes locally?
-   [x] Have you created a changelog file for each project being changed, ie `pnpm changelog add --filter=<project>`?

<!-- Mark completed items with an [x] -->

### FOR PR REVIEWER ONLY:

-   [ ] I have reviewed that everything is sanitized/escaped appropriately for any SQL or XSS injection possibilities. I made sure Linting is not ignored or disabled.

joshuatf committed:
Add initial ImageGallery component

joshuatf committed:
Add changelog entry

joshuatf committed:
Extract list item component and styles

joshuatf committed:
Rename SortableList to Sortable

joshuatf committed:
Add placeholders for dropping items

joshuatf committed:
Fix drag throttling delays

joshuatf committed:
Use placeholder component

joshuatf committed:
Move handles back to sortable

joshuatf committed:
Add horizontal option to sortable

joshuatf committed:
Use CSS for placeholder styles

joshuatf committed:
Remove placeholder component

joshuatf committed:
Add image gallery styles

joshuatf committed:
Add last droppable class

joshuatf committed:
Extract class utils and add tests

joshuatf committed:
Add list item stories

joshuatf committed:
Add readmes

octaedro commented on packages/js/components/src/image-gallery/image-gallery-item.scss:
```diff
@@ -0,0 +1,22 @@
+    }
+
+    .woocommerce-pill {
+        background: #007CBA;
```
I wonder if we could use something like `background: var(--wp-admin-theme-color);` instead of fixing the color.

joshuatf replied:
Definitely!  Good call.

octaedro commented on packages/js/components/src/sortable/sortable-item.scss:
```diff
@@ -0,0 +1,54 @@
+        display: none;
+        height: $place-holder-size;
+        width: 100%;
+        background: #007CBA;
```
Same here

octaedro commented on packages/js/components/src/image-gallery/image-gallery.tsx:
```diff
@@ -0,0 +1,34 @@
+/**
+ * External dependencies
+ */
+import { Children, createElement, Fragment } from '@wordpress/element';
```
I think that the `Fragment` is not being used here.

octaedro commented on packages/js/components/src/list-item/stories/index.tsx:
```diff
@@ -0,0 +1,43 @@
+/**
+ * External dependencies
+ */
+import { createElement, Fragment } from '@wordpress/element';
```
This `Fragment` is not being used either.

octaedro commented on packages/js/components/src/sortable/README.md:
```diff
@@ -20,8 +20,8 @@ This component accepts any valid JSX elements as children.  Adding a `key` to el
 Name | Type | Default | Description
```
We should also update the `title` and the `usage` areas to reflect the new component name.

joshuatf replied:
Good catch!  Updated.

octaedro reviewed (commented):
You did a great job here @joshuatf. The code looks mainly good (I just left a couple of small suggestions). I experienced a couple of problems while testing this PR using the storybook.

Trying to use the drag and drop:
- it's not working in `ListItem > Draggable`.
- in `ImageGallery`, the drag and drop looks a little bit choppy when I took an item, and moved it to the last position.

joshuatf committed:
Handle PR feedback

joshuatf commented:
Thanks for the review, @octaedro!  Addressed most of your feedback.

>it's not working in ListItem > Draggable.

I had intentionally not made this draggable just to display how draggable handles are added, but I see how that's a bit confusing.  Went ahead and added a `Sortable` wrapper around this example.

>in ImageGallery, the drag and drop looks a little bit choppy when I took an item, and moved it to the last position.

Could you capture a video of this?  We might want to save animation/performance related stuff for a follow-up to unblock the issues in the current cycle, but I'm happy to add it here if it's a quick fix.

octaedro commented:
> Could you capture a video of this? We might want to save animation/performance related stuff for a follow-up to unblock the issues in the current cycle, but I'm happy to add it here if it's a quick fix.

I'm still experiencing the drop issue. As you can see in the gif below, it takes something like a second to put the image in the right place (if I move an image to the last position). I tried it in Chrome and Brave.

![Screen Capture on 2022-09-01 at 10-47-50](https://user-images.githubusercontent.com/1314156/187931640-29d50122-35e4-42cc-998f-cdf2b8b372e0.gif)

Maybe it's easier to see here:
![Screen Capture on 2022-09-01 at 11-09-47](https://user-images.githubusercontent.com/1314156/187935090-6b0250b0-3a39-4238-a0b3-8cd3b78152ec.gif)

Since I was able to see during the demo that it's working well on your end, we can merge this PR and if the image delay is something that is not just happening on my end, tackle that in a follow-up PR.

octaedro reviewed (dismissed):
Thank you @joshuatf for addressing the changes. LGTM 🚀

octaedro reviewed (approved):
LGTM :shipit:
